File: simplevlcplayer/src/main/java/com/masterwok/simplevlcplayer/fragments/BasePlayerFragment.java
package com.masterwok.simplevlcplayer.fragments;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.net.Uri;
import android.os.Bundle;
import android.os.IBinder;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentManager;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.support.v4.provider.DocumentFile;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ProgressBar;

import com.masterwok.simplevlcplayer.R;
import com.masterwok.simplevlcplayer.common.utils.ResourceUtil;
import com.masterwok.simplevlcplayer.common.utils.ThreadUtil;
import com.masterwok.simplevlcplayer.common.utils.ViewUtil;
import com.masterwok.simplevlcplayer.components.PlayerControlComponent;
import com.masterwok.simplevlcplayer.contracts.MediaPlayer;
import com.masterwok.simplevlcplayer.dagger.injectors.InjectableFragment;
import com.masterwok.simplevlcplayer.services.MediaPlayerService;
import com.masterwok.simplevlcplayer.services.binders.MediaPlayerServiceBinder;


public abstract class BasePlayerFragment
        extends InjectableFragment
        implements PlayerControlComponent.Callback
        , MediaPlayer.Callback {

    public static final String MediaUri = "extra.mediauri";
    public static final String SubtitleUri = "extra.subtitleuri";
    public static final String OpenSubtitlesUserAgent = "extra.useragent";
    public static final String SubtitleDestinationUri = "extra.subtitledestinationuri";
    public static final String SubtitleLanguageCode = "extra.subtitlelanguagecode";

    protected MediaPlayerServiceBinder serviceBinder;
    private Uri subtitleDestinationUri;
    protected Uri subtitleUri;
    protected Uri mediaUri;
    private String openSubtitlesUserAgent;
    private String subtitleLanguageCode;

    private MediaControllerCompat mediaController;
    private ProgressBar progressBar;

    protected abstract void configure(
            boolean isPlaying,
            long time,
            long length
    );

    protected abstract void onConnected();

    protected abstract void onDisconnected();

    private ServiceConnection mediaPlayerServiceConnection = new ServiceConnection() {

        @Override
        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
            BasePlayerFragment.this.serviceBinder = (MediaPlayerServiceBinder) iBinder;

            serviceBinder.setCallback(BasePlayerFragment.this);

            onConnected();

            registerMediaController(serviceBinder);
        }


        @Override
        public void onServiceDisconnected(ComponentName componentName) {
            serviceBinder = null;

            onDisconnected();
        }
    };

    private MediaControllerCompat.Callback controllerCallback = new MediaControllerCompat.Callback() {
        @Override
        public void onPlaybackStateChanged(PlaybackStateCompat state) {
            configure(
                    state.getState() == PlaybackStateCompat.STATE_PLAYING,
                    state.getPosition(),
                    state.getBufferedPosition()
            );
        }
    };

    @Override
    public void onStart() {
        super.onStart();

        bindMediaPlayerService();
    }

    @Override
    public void onStop() {
        serviceBinder.setCallback(null);
        mediaController.unregisterCallback(controllerCallback);
        unbindMediaPlayerService();

        super.onStop();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        readIntent();
    }

    private void readIntent() {
        Intent intent = getActivity().getIntent();

        mediaUri = intent.getParcelableExtra(MediaUri);
        subtitleUri = intent.getParcelableExtra(SubtitleUri);
        subtitleDestinationUri = intent.getParcelableExtra(SubtitleDestinationUri);
        openSubtitlesUserAgent = intent.getStringExtra(OpenSubtitlesUserAgent);
        subtitleLanguageCode = intent.getStringExtra(SubtitleLanguageCode);
    }

    @Override
    public void onViewCreated(
            @NonNull View view,
            @Nullable Bundle savedInstanceState
    ) {
        super.onViewCreated(view, savedInstanceState);

        initProgressBar();
    }

    private void unbindMediaPlayerService() {
        final Activity activity = getActivity();

        if (activity == null) {
            return;
        }

        activity.unbindService(mediaPlayerServiceConnection);
        serviceBinder = null;
    }

    private void bindMediaPlayerService() {
        getActivity().bindService(
                getMediaPlayerServiceIntent(),
                mediaPlayerServiceConnection,
                Context.BIND_AUTO_CREATE
        );
    }

    private Intent getMediaPlayerServiceIntent() {
        return new Intent(
                getActivity(),
                MediaPlayerService.class
        );
    }

    private void registerMediaController(MediaPlayerServiceBinder serviceBinder) {
        final Activity activity = getActivity();

        if (activity == null || serviceBinder == null) {
            return;
        }

        mediaController = new MediaControllerCompat(
                activity,
                serviceBinder.getMediaSession()
        );

        mediaController.registerCallback(controllerCallback);

        MediaControllerCompat.setMediaController(activity, mediaController);
    }

    @Override
    public void onPlayPauseButtonClicked() {
        if (serviceBinder == null) {
            return;
        }

        serviceBinder.togglePlayback();
    }

    @Override
    public void onCastButtonClicked() {
        FragmentManager fragmentManager = getFragmentManager();

        if (fragmentManager == null) {
            return;
        }

        new RendererItemDialogFragment().show(
                fragmentManager,
                RendererItemDialogFragment.Tag
        );
    }

    @Override
    public void onSubtitlesButtonClicked() {
        FragmentManager fragmentManager = getFragmentManager();

        if (fragmentManager == null) {
            return;
        }

        String mediaName = DocumentFile
                .fromSingleUri(getContext(), mediaUri)
                .getName();

        SubtitlesDialogFragment
                .createInstance(
                        mediaName
                        , openSubtitlesUserAgent
                        , subtitleLanguageCode
                        , subtitleDestinationUri
                ).show(fragmentManager, SubtitlesDialogFragment.Tag);
    }

    @Override
    public void onProgressChangeStarted() {
        if (serviceBinder == null) {
            return;
        }

        serviceBinder.pause();
    }

    @Override
    public void onProgressChanged(int progress) {
        if (serviceBinder == null) {
            return;
        }

        serviceBinder.setProgress(progress);
        serviceBinder.play();
    }

    @SuppressWarnings("ConstantConditions")
    private void initProgressBar() {
        final Context context = getContext();

        progressBar = new ProgressBar(
                context,
                null,
                android.R.attr.progressBarStyleLarge
        );

        progressBar.setVisibility(View.GONE);

        ViewUtil.setProgressBarColor(
                context,
                progressBar,
                R.color.progress_bar_spinner
        );

        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
                ResourceUtil.getDimenDp(context, R.dimen.player_spinner_width),
                ResourceUtil.getDimenDp(context, R.dimen.player_spinner_height)
        );

        params.gravity = Gravity.CENTER;

        ((ViewGroup) getView()).addView(progressBar, params);
    }

    @Override
    public void onPlayerOpening() {
    }

    @Override
    public void onBuffering(float buffering) {
        if (buffering == 100f) {
            ThreadUtil.onMain(() -> progressBar.setVisibility(View.GONE));
            return;
        }

        if (progressBar.getVisibility() == View.VISIBLE) {
            return;
        }

        ThreadUtil.onMain(() -> progressBar.setVisibility(View.VISIBLE));
    }

    @Override
    public void onPlayerSeekStateChange(boolean canSeek) {
    }

    @Override
    public void onPlayerPlaying() {
    }

    @Override
    public void onPlayerPaused() {
    }

    @Override
    public void onPlayerStopped() {
    }

    @Override
    public void onPlayerEndReached() {
        //noinspection ConstantConditions
        getActivity().finish();
    }

    @Override
    public void onPlayerError() {
    }

    @Override
    public void onPlayerTimeChange(long timeChanged) {
    }

    @Override
    public void onPlayerPositionChanged(float positionChanged) {
    }

    @Override
    public void onSubtitlesCleared() {
    }

}

File: simplevlcplayer/src/main/java/com/masterwok/simplevlcplayer/fragments/BasePlayerFragment.kt
package com.masterwok.simplevlcplayer.fragments

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.net.Uri
import android.os.Bundle
import android.os.IBinder
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.support.v4.provider.DocumentFile
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.webkit.URLUtil
import android.widget.FrameLayout
import android.widget.ProgressBar
import com.masterwok.simplevlcplayer.R
import com.masterwok.simplevlcplayer.common.AndroidJob
import com.masterwok.simplevlcplayer.common.extensions.setColor
import com.masterwok.simplevlcplayer.common.utils.ResourceUtil
import com.masterwok.simplevlcplayer.components.PlayerControlComponent
import com.masterwok.simplevlcplayer.contracts.MediaPlayer
import com.masterwok.simplevlcplayer.dagger.injectors.InjectableFragment
import com.masterwok.simplevlcplayer.services.MediaPlayerService
import com.masterwok.simplevlcplayer.services.binders.MediaPlayerServiceBinder
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.launch

abstract class BasePlayerFragment : InjectableFragment()
        , PlayerControlComponent.Callback
        , MediaPlayer.Callback {

    companion object {
        const val MediaUri = "extra.mediauri"
        const val SubtitleUri = "extra.subtitleuri"
        const val OpenSubtitlesUserAgent = "extra.useragent"
        const val SubtitleDestinationUri = "extra.subtitledestinationuri"
        const val SubtitleLanguageCode = "extra.subtitlelanguagecode"
    }

    protected var serviceBinder: MediaPlayerServiceBinder? = null

    private var mediaController: MediaControllerCompat? = null
    private var subtitleDestinationUri: Uri? = null
    private var openSubtitlesUserAgent: String? = null
    private var subtitleLanguageCode: String? = null

    protected var subtitleUri: Uri? = null
    protected var mediaUri: Uri? = null

    private lateinit var progressBar: ProgressBar

    private val rootJob: AndroidJob = AndroidJob(lifecycle)

    protected abstract fun configure(
            isPlaying: Boolean
            , time: Long
            , length: Long
    )

    protected abstract fun onConnected()

    protected abstract fun onDisconnected()

    private val mediaPlayerServiceConnection = object : ServiceConnection {

        override fun onServiceConnected(componentName: ComponentName, iBinder: IBinder) {
            this@BasePlayerFragment.serviceBinder = iBinder as MediaPlayerServiceBinder

            serviceBinder?.callback = this@BasePlayerFragment

            onConnected()

            registerMediaController(serviceBinder)
        }


        override fun onServiceDisconnected(componentName: ComponentName) {
            serviceBinder = null

            onDisconnected()
        }
    }

    private val controllerCallback = object : MediaControllerCompat.Callback() {
        override fun onPlaybackStateChanged(state: PlaybackStateCompat) {
            configure(
                    state.state == PlaybackStateCompat.STATE_PLAYING,
                    state.position,
                    state.bufferedPosition
            )
        }
    }

    private fun registerMediaController(serviceBinder: MediaPlayerServiceBinder?) {
        val activity = activity

        if (activity == null || serviceBinder == null) {
            return
        }

        mediaController = MediaControllerCompat(
                activity,
                serviceBinder.mediaSession!!
        ).apply {
            registerCallback(controllerCallback)
        }

        MediaControllerCompat.setMediaController(activity, mediaController)
    }

    override fun onStart() {
        super.onStart()

        bindMediaPlayerService()
    }

    override fun onStop() {
        serviceBinder?.callback = null
        mediaController?.unregisterCallback(controllerCallback)
        unbindMediaPlayerService()

        super.onStop()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)

        readIntent()
    }

    private fun readIntent() {
        val intent = activity?.intent

        mediaUri = intent?.getParcelableExtra(MediaUri)
        subtitleUri = intent?.getParcelableExtra(SubtitleUri);
        subtitleDestinationUri = intent?.getParcelableExtra(SubtitleDestinationUri);
        openSubtitlesUserAgent = intent?.getStringExtra(OpenSubtitlesUserAgent);
        subtitleLanguageCode = intent?.getStringExtra(SubtitleLanguageCode);
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        initProgressBar()
    }

    private fun initProgressBar() {
        progressBar = ProgressBar(
                context
                , null
                , android.R.attr.progressBarStyleLarge
        ).apply {
            visibility = View.GONE
            setColor(R.color.progress_bar_spinner)
        }

        val params = FrameLayout.LayoutParams(
                ResourceUtil.getDimenDp(context!!, R.dimen.player_spinner_width),
                ResourceUtil.getDimenDp(context!!, R.dimen.player_spinner_height)
        ).apply {
            gravity = Gravity.CENTER
        }

        (view as ViewGroup).addView(
                progressBar
                , params
        )
    }

    private fun unbindMediaPlayerService() {
        activity?.unbindService(mediaPlayerServiceConnection)
        serviceBinder = null
    }

    private fun bindMediaPlayerService() = activity?.bindService(
            getMediaPlayerIntent()
            , mediaPlayerServiceConnection
            , Context.BIND_AUTO_CREATE
    )

    private fun getMediaPlayerIntent(): Intent? {
        if (activity == null) {
            return null
        }

        return Intent(activity, MediaPlayerService::class.java)
    }


    override fun onPlayPauseButtonClicked() {
        serviceBinder?.togglePlayback()
    }

    override fun onCastButtonClicked() = RendererItemDialogFragment().show(
            fragmentManager,
            RendererItemDialogFragment.Tag
    )


    override fun onSubtitlesButtonClicked() {
        val fragmentManager = fragmentManager ?: return

        val mediaName = if (URLUtil.isContentUrl(mediaUri.toString()))
            DocumentFile.fromSingleUri(context, mediaUri).name
        else {
            mediaUri?.lastPathSegment
        }

        SubtitlesDialogFragment.createInstance(
                mediaName!!
                , subtitleUri
                , openSubtitlesUserAgent
                , subtitleLanguageCode
                , subtitleDestinationUri
        ).show(fragmentManager, SubtitlesDialogFragment.Tag)
    }

    override fun onProgressChangeStarted() {
        serviceBinder?.pause()
    }

    override fun onProgressChanged(progress: Int) {
        serviceBinder?.setProgress(progress)
        serviceBinder?.play()
    }

    override fun onPlayerOpening() {}

    override fun onBuffering(buffering: Float) {
        if (buffering == 100f) {
            launch(UI, parent = rootJob) { progressBar.visibility = View.GONE }
            return
        }

        if (progressBar.visibility == View.VISIBLE) {
            return
        }

        launch(UI, parent = rootJob) { progressBar.visibility = View.VISIBLE }
    }

    override fun onPlayerSeekStateChange(canSeek: Boolean) {}

    override fun onPlayerPlaying() {}

    override fun onPlayerPaused() {}

    override fun onPlayerStopped() {}

    override fun onPlayerEndReached() {
        activity?.finish()
    }

    override fun onPlayerError() {}

    override fun onPlayerTimeChange(timeChanged: Long) {}

    override fun onPlayerPositionChanged(positionChanged: Float) {}

    override fun onSubtitlesCleared() {}
}